### PR TITLE
[Core] Backport std::integer_sequence & co. from libstdc++

### PR DESCRIPTION
**Pull request by bluehood**
~~or, if __cplusplus > 201103L, just #include <utility>~~
or just include `<utility>` if `std::integer_sequence` is available.

### CODE REVIEW

File: core/foundation/inc/ROOT/RIntegerSequence.hxx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#ifndef ROOT_RIntegerSequence
+#define ROOT_RIntegerSequence
+
+#if __cplusplus > 201103L
+// no need to backport
+
+#include <utility>
+
+#else
+// backport libstdc++'s implementation (gcc7.3.1)
```
Comment (pcanal) [inline core/foundation/inc/ROOT/RIntegerSequence.hxx]: Please add link of where you got it from if possible (and if from git either a commit number here or in our commit message) .. i.e. make it as easy as possible to find potential updates :)

Comment (bluehood) [reply core/foundation/inc/ROOT/RIntegerSequence.hxx]: I can't be more precise than "libstdc++'s implementation shipped with gcc 7.3.1", but probably I can dig up a link to the `utility` header for this version of gcc on github.

The other backports would also benefit from such links probably :sweat_smile:

File: core/foundation/inc/ROOT/RIntegerSequence.hxx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RIntegerSequence
+#define ROOT_RIntegerSequence
+
+#if __cplusplus > 201103L
```
[the remark this answers is not in the document]
Comment (bluehood) [reply core/foundation/inc/ROOT/RIntegerSequence.hxx]: I see, what's the correct `#if` condition?

`rhysd_array_view.hxx`, `RNotFn.hxx` and `RMakeUnique.hxx` also rely on `__cplusplus` by the way

[the remark this answers is not in the document]
Comment (pcanal) [reply core/foundation/inc/ROOT/RIntegerSequence.hxx]: See core/foundation/inc/ROOT/RStringView.hxx (the implementation copy is in a separate file which contains the link but is frustratingly missing the svn revision :( ).  

> rhysd_array_view.hxx, RNotFn.hxx and RMakeUnique.hxx also rely on __cplusplus by the way

So we have been lucky-ish so far (But then again maybe we are not building with hardly compliant compilers like icc) and indeed it is a development tradeoff/bet as doing the 'right' thing require writing a small feature test to add in CMakeFile.txt and adding a configurable macro to the config file vs having to do that later when we encounter a stupid compiler ....

File: core/foundation/inc/ROOT/RIntegerSequence.hxx (hunk 1 above)
Comment (bluehood) [reply core/foundation/inc/ROOT/RIntegerSequence.hxx]: link added in the line below

File: core/foundation/inc/ROOT/RIntegerSequence.hxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (bluehood) [reply core/foundation/inc/ROOT/RIntegerSequence.hxx]: ok, I added the check -- see second commit in the PR

File: cmake/modules/RootConfiguration.cmake [hunk 3]
```diff
@@ -599,6 +599,14 @@ else()
   set(hasstdinvoke undef)
 endif()
 
+CHECK_CXX_SOURCE_COMPILES("#include <utility>
+int main() { std::make_index_sequence<3> s; return 0; }" found_stdindexsequence)
```
Comment (pcanal) [inline cmake/modules/RootConfiguration.cmake]: Should we test getting the value (i.e. making minimally sure the 'api' is the one expected?

Comment (bluehood) [reply cmake/modules/RootConfiguration.cmake]: Are you saying there are standard library implementations with a `std::make_index_sequence` implementation that _does not_ return the expected `std::index_sequence` and that we want to support them anyway?

At that point nothing is real anymore :sweat_smile:

I can make it
```c++
CHECK_CXX_SOURCE_COMPILES("#include <utility>
#include <type_traits>
int main() { static_assert(std::is_same<std::integer_sequence<std::size_t, 0, 1, 1>, std::make_index_sequence<3>>::value, ""); return 0; }"
```
would that be ok?

Comment (pcanal) [reply cmake/modules/RootConfiguration.cmake]: Looks fine.  Probably even more than I asked :).

Comment (bluehood) [reply cmake/modules/RootConfiguration.cmake]: I wasn't sure what you meant with "test getting the value".
I added the check.

**Comment by bluehood** [comment]
Hi, all is green and all comments have been addressed, can I merge?

**Comment by Axel-Naumann** [comment]
Yes please merge!

**Pull request merged**